arrow/formatter.py

```diff
@@ -22,19 +31,24 @@ class DateTimeFormatter:
     # an atomic group. For more info on atomic groups and how to they are
     # emulated in Python's re library, see https://stackoverflow.com/a/13577411/2701578
 
-    _FORMAT_RE = re.compile(
+    _FORMAT_RE: Final[Pattern[str]] = re.compile(
         r"(\[(?:(?=(?P<literal>[^]]))(?P=literal))*\]|YYY?Y?|MM?M?M?|Do|DD?D?D?|d?dd?d?|HH?|hh?|mm?|ss?|SS?S?S?S?S?|ZZ?Z?|a|A|X|x|W)"
     )
 
-    def __init__(self, locale="en_us"):
+    locale: locales.Locale
+
+    def __init__(self, locale: str = "en_us") -> None:
 
         self.locale = locales.get_locale(locale)
 
-    def format(cls, dt, fmt):
+    def format(cls, dt: datetime, fmt: str) -> str:
 
-        return cls._FORMAT_RE.sub(lambda m: cls._format_token(dt, m.group(0)), fmt)
+        # FIXME: _format_token() is nullable
+        return cls._FORMAT_RE.sub(
+            lambda m: cast(str, cls._format_token(dt, m.group(0))), fmt
+        )
 
-    def _format_token(self, dt, token):
+    def _format_token(self, dt: datetime, token: Optional[str]) -> Optional[str]:
```

Comment:
In what case can `_format_token` return `None`?

Reply:
@jadchaar  If token is `None` or any `str` that does not match with if conditions.

Reply:
Could these FIXMEs, if left unresolved, cause issues for users in regular use? When will the nullable issues you mark with the FIXME label cause problems?

Reply:
It could be. but because I didn't change actual behavior of `formatter.py`, this PR will not introduce new error to `formatter.py`.
BTW, I can make `token` as `str` because [`Match.group()`](https://docs.python.org/3/library/re.html#re.Match.group) always return `AnyStr`.

Reply:
If we can assure that `token` always match one branch in `_format_token`, I have to narrow down type of `token` to some `Literal`.

Reply:
To test things out, I put a `raise ValueError` at the bottom of the control statements to see if I could possibly get anything to fall through, but I was unable to ever reach beyond the conditionals.

I tried these:

```python
now = arrow.now()

now.format()
now.format("")
now.format(None)
```

The only thing that caused trouble was the final None one, but it failed on the call to `sub` before even going into `_format_token`. I cannot seem to find a case that digs all the way down to the bottom of the `_format_token` function, so we might be safe to say it always returns a string and never `None`.

Reply:
When I run the code with the `raise ValueError` at the end of `_format_token`, the only test that fails in the test suite is `test_nonsense` in `test_formatter.py`:


```console
====================================================================================================================================================== FAILURES ======================================================================================================================================================
_______________________________________________________________________________________________________________________________________ TestFormatterFormatToken.test_nonsense _______________________________________________________________________________________________________________________________________

self = <tests.test_formatter.TestFormatterFormatToken object at 0x1034c1d00>

    def test_nonsense(self):
        dt = datetime(2012, 1, 1, 11)
>       assert self.formatter._format_token(dt, None) is None

tests/test_formatter.py:166: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <arrow.formatter.DateTimeFormatter object at 0x1025cf670>, dt = datetime.datetime(2012, 1, 1, 11, 0), token = None

    def _format_token(self, dt: datetime, token: Optional[str]) -> Optional[str]:
    
        if token and token.startswith("[") and token.endswith("]"):
            return token[1:-1]
    
        if token == "YYYY":
            return self.locale.year_full(dt.year)
        if token == "YY":
            return self.locale.year_abbreviation(dt.year)
    
        if token == "MMMM":
            return self.locale.month_name(dt.month)
        if token == "MMM":
            return self.locale.month_abbreviation(dt.month)
        if token == "MM":
            return f"{dt.month:02d}"
        if token == "M":
            return f"{dt.month}"
    
        if token == "DDDD":
            return f"{dt.timetuple().tm_yday:03d}"
        if token == "DDD":
            return f"{dt.timetuple().tm_yday}"
        if token == "DD":
            return f"{dt.day:02d}"
        if token == "D":
            return f"{dt.day}"
    
        if token == "Do":
            return self.locale.ordinal_number(dt.day)
    
        if token == "dddd":
            return self.locale.day_name(dt.isoweekday())
        if token == "ddd":
            return self.locale.day_abbreviation(dt.isoweekday())
        if token == "d":
            return f"{dt.isoweekday()}"
    
        if token == "HH":
            return f"{dt.hour:02d}"
        if token == "H":
            return f"{dt.hour}"
        if token == "hh":
            return f"{dt.hour if 0 < dt.hour < 13 else abs(dt.hour - 12):02d}"
        if token == "h":
            return f"{dt.hour if 0 < dt.hour < 13 else abs(dt.hour - 12)}"
    
        if token == "mm":
            return f"{dt.minute:02d}"
        if token == "m":
            return f"{dt.minute}"
    
        if token == "ss":
            return f"{dt.second:02d}"
        if token == "s":
            return f"{dt.second}"
    
        if token == "SSSSSS":
            return f"{dt.microsecond:06d}"
        if token == "SSSSS":
            return f"{dt.microsecond // 10:05d}"
        if token == "SSSS":
            return f"{dt.microsecond // 100:04d}"
        if token == "SSS":
            return f"{dt.microsecond // 1000:03d}"
        if token == "SS":
            return f"{dt.microsecond // 10000:02d}"
        if token == "S":
            return f"{dt.microsecond // 100000}"
    
        if token == "X":
            return f"{dt.timestamp()}"
    
        if token == "x":
            return f"{dt.timestamp() * 1_000_000:.0f}"
    
        if token == "ZZZ":
            return dt.tzname()
    
        if token in ["ZZ", "Z"]:
            separator = ":" if token == "ZZ" else ""
            tz = dateutil_tz.tzutc() if dt.tzinfo is None else dt.tzinfo
            # `dt` must be aware object. Otherwise, this line will raise AttributeError
            # https://github.com/arrow-py/arrow/pull/883#discussion_r529866834
            # datetime awareness: https://docs.python.org/3/library/datetime.html#aware-and-naive-objects
            total_minutes = int(cast(timedelta, tz.utcoffset(dt)).total_seconds() / 60)
    
            sign = "+" if total_minutes >= 0 else "-"
            total_minutes = abs(total_minutes)
            hour, minute = divmod(total_minutes, 60)
    
            return f"{sign}{hour:02d}{separator}{minute:02d}"
    
        if token in ("a", "A"):
            return self.locale.meridian(dt.hour, token)
    
        if token == "W":
            year, week, day = dt.isocalendar()
            return f"{year}-W{week:02d}-{day}"
    
    
>       raise ValueError("FOO BAR")
E       ValueError: FOO BAR

arrow/formatter.py:150: ValueError
```

I don't know when `sub` would ever pass `None` to` _format_token` though... this test case is directly forcing `None` into the `_format_token` function.

---

@isac322 I think a possible solution here could be to simply put a final case in the fall through to just return the empty string. This will take care of all these cases that do not match any of the conditionals in a sane way. It will also allow us to make the return type `str` rather than `Optional[str]`. Thoughts?

Reply:
@jadchaar Sounds good! I will add final return statement and remove `Optional`.

Reply:
@jadchaar mypy still complains about `Optional`.
```
arrow/formatter.py:125: error: Incompatible return value type (got "Optional[str]", expected "str")
arrow/formatter.py:142: error: Incompatible return value type (got "Optional[str]", expected "str")
```

First one is `return dt.tzname()` case.
I know that `Arrow` always store non naive `datetime` now. but because `DateTimeFormatter` is public, someone could pass naive `datetime` object to `DateTimeFormatter.format()` and in that case `DateTimeFormatter._format_token()` will return `None`.

Second one is `return self.locale.meridian(dt.hour, token)`.


But because those issues are known issue (I only annotate type for those cases), I think we can just leave FIXME comment and cast it as `str` for now. I want to minimize behavior changing of this PR.

Reply:
Given Isac's latest findings, you think this is good to merge @systemcatch? We might want to consider creating some issues to address these two FIXMEs going forward (might require some significant refactoring) if possible.

Reply:
Hey @jadchaar I created #918 to track these 2 bugs. We should merge this in now (exciting!).